### PR TITLE
Doc symtable

### PR DESCRIPTION
**Pull request by lanarimarco**
## Checklist:
- [ ] There are tests regarding this feature
- [X] The code follows the Kotlin conventions (run `./gradlew ktlintCheck`)
- [X] The code passes all tests (run `./gradlew check`)
- [X] There is a specific documentation in the `docs` directory

### CODE REVIEW

File: docs/symboltable_serialization.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Abstract
+In this document we are going to present how Jariko allows managing programs memory persistence using as approach the serialization.
```
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:2]: "explain" instead of "present"

File: docs/symboltable_serialization.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
+# Abstract
+In this document we are going to present how Jariko allows managing programs memory persistence using as approach the serialization.  
+As everyone knows the symbol table is an object containing the program memory, and generally speaking
```
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:3]: i would remove "As everyone knows"

File: docs/symboltable_serialization.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+it is nothing more than a `Map<String, Value>` with String: the variable name, and Value: the value of it.  
+Symbol table has scope program and lives within the program, but when the program dies, if that program runs into an 
+activation group, we need to recover its state, for this reason we have thought about serialization.
```
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:6]: not "if that program runs into an activation group" but "if that program finishes with RT indicator"

File: docs/symboltable_serialization.md (hunk 1 above)
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:2]: not "how Jariko allows managing programs memory persistence using as approach the ." but "how Jariko uses serialization to manage programs' memory's persistence"

File: docs/symboltable_serialization.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+The following chart shows the objects involved with particular relevance to what happens over the time.  
+We suppose that Jariko will be run three different programs in the same thread.  
+The almost hidden objects (gray) are *not active*, the others are *active*, but all objects *are living*
+in the same instance of Jariko.
+
+![PD000236-SerializationModel_1.png](images/PD000236-SerializationModel_1.png)
+
+## InternalInterpreter
+For each interpreted PGM we have an instance of InternalInterpreter having its own instance of SymbolTable.  
+You can think about the follow calls stack: PGM1 -> PGM2 -> PGMi and as you can guess, PGMi is the current execution program.
```
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:18]: not "is the current execution program" but "is the program currently executed"

File: docs/symboltable_serialization.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+in the same instance of Jariko.
+
+![PD000236-SerializationModel_1.png](images/PD000236-SerializationModel_1.png)
+
+## InternalInterpreter
+For each interpreted PGM we have an instance of InternalInterpreter having its own instance of SymbolTable.  
+You can think about the follow calls stack: PGM1 -> PGM2 -> PGMi and as you can guess, PGMi is the current execution program.  
+
+## MemorySlice
+This object tells us that the SymbolTable and its own persistence are different things. The MemorySlice is nothing
```
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:21]: not "The MemorySlice is nothing of more complex than an association between" but "The MemorySlice is simply the association between"

File: docs/symboltable_serialization.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+This object tells us that the SymbolTable and its own persistence are different things. The MemorySlice is nothing 
+of more complex than an association between the SymbolTable instance and the MemorySliceId.  
+MemorySliceId is identified just by following properties:
+ * Program
+ * Activation group  
+
+In conclusion, we can say that we will have an instance of MemorySlice for each SymbolTable.
+
+## MemorySliceMgr
+This object, behind the scenes, handles serialization and de-serialization mechanisms.
```
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:30]: not "behind the scenes" but "under the hood"

File: docs/symboltable_serialization.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+### Deserialization
+For each SymbolTable belonging to a program running within an activation group, the initialization phase includes:
+ * The creation of MemorySlice in which we couple SymbolTable with MemorySliceId
```
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:34]: not "couple" but "associate"

File: docs/symboltable_serialization.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ * The reading from IMemorySliceStorage and retrieving of `Map<String, Value>` containing stored values associated to MemorySliceId
+ * The SymbolTable initialization using `Map<String, Value>` returned by the previous step
+
+### Serialization
+The serialization is triggered only at the end of the main program execution.  
+For each MemorySlice having property `persist=true` (program running into activation group), MemorySlice will be passed to
```
Comment (benetti-smeup) [inline docs/symboltable_serialization.md:40]: not "program running into activation group" but "program finishing with RT indicator"

**Review by benetti-smeup — CHANGES_REQUESTED** [review]
I added some comments

File: docs/symboltable_serialization.md (hunk 1 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md:2]: ok

File: docs/symboltable_serialization.md (hunk 2 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md:3]: ok

File: docs/symboltable_serialization.md (hunk 3 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md:6]: ok

File: docs/symboltable_serialization.md (hunk 1 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md:2]: ok

File: docs/symboltable_serialization.md (hunk 4 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md:18]: ok

File: docs/symboltable_serialization.md (hunk 5 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md:21]: ok

File: docs/symboltable_serialization.md (hunk 7 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md:34]: ok

File: docs/symboltable_serialization.md (hunk 8 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md:40]: ok

File: docs/symboltable_serialization.md (hunk 6 above)
Comment (lanarimarco) [reply docs/symboltable_serialization.md]: ok

**Review by benetti-smeup — APPROVED** [review]
ok

**Pull request merged**